File: frontend/features/battle/mechanics/damage.ts
import type { BattleMonster } from "../types";
import type { BattleType } from "../data/types";
import { getModifiedStat } from "./statStage";
import { getTypeMultiplier } from "./typeChart";

type CalculateDamageParams = {
  attacker: BattleMonster;
  defender: BattleMonster;
  power: number;
  moveType: BattleType;
};

export type DamageResult = {
  damage: number;
  critical: boolean;
};

function getBurnModifiedAttack(
  attack: number,
  attacker: BattleMonster
): number {
  if (attacker.mainStatus?.condition !== "burn") {
    return attack;
  }

  return Math.max(1, Math.floor(attack / 2));
}

function getStabMultiplier(
  attacker: BattleMonster,
  moveType: BattleType
): number {
  return attacker.type === moveType ? 1.5 : 1;
}

function isCriticalHit(): boolean {
  return Math.random() < 1 / 16;
}

function getCriticalMultiplier(critical: boolean): number {
  return critical ? 1.5 : 1;
}

function getRandomMultiplier(): number {
  return 0.85 + Math.random() * 0.15;
}

export function calculateDamage({
  attacker,
  defender,
  power,
  moveType,
}: CalculateDamageParams): DamageResult {
  if (power <= 0) {
    return {
      damage: 0,
      critical: false,
    };
  }

  const modifiedAttack = getModifiedStat(
    attacker.attack,
    attacker.statStages.attack
  );

  const attack = getBurnModifiedAttack(
    modifiedAttack,
    attacker
  );

  const defense = getModifiedStat(
    defender.defense,
    defender.statStages.defense
  );

  const typeMultiplier = getTypeMultiplier(
    moveType,
    defender.type
  );

  const stabMultiplier = getStabMultiplier(
    attacker,
    moveType
  );

  const critical = isCriticalHit();
  const criticalMultiplier = getCriticalMultiplier(critical);
  const randomMultiplier = getRandomMultiplier();

  const baseDamage = power + attack - defense;
  const damage = Math.max(1, baseDamage);

  return {
    damage: Math.max(
      1,
      Math.floor(
        damage *
          typeMultiplier *
          stabMultiplier *
          criticalMultiplier *
          randomMultiplier
      )
    ),
    critical,
  };
}

File: backend/internal/battle/damage.go
package battle

import (
	"math"
	"math/rand"
)

// DamageResult はダメージ計算の結果。
type DamageResult struct {
	Damage   int
	Critical bool
}

func getBurnModifiedAttack(attack int, attacker BattleMonster) int {
	if attacker.MainStatus == nil || attacker.MainStatus.Condition != StatusBurn {
		return attack
	}

	return maxInt(1, attack/2)
}

func getStabMultiplier(attacker BattleMonster, moveType BattleType) float64 {
	if attacker.Type == moveType {
		return 1.5
	}
	return 1
}

func isCriticalHit() bool {
	return rand.Float64() < 1.0/16
}

func getCriticalMultiplier(critical bool) float64 {
	if critical {
		return 1.5
	}
	return 1
}

func getRandomMultiplier() float64 {
	return 0.85 + rand.Float64()*0.15
}

// CalculateDamageParams はCalculateDamageへの入力。
type CalculateDamageParams struct {
	Attacker BattleMonster
	Defender BattleMonster
	Power    int
	MoveType BattleType
}

// CalculateDamage はfeatures/battle/mechanics/damage.tsの移植。
// ダメージ = (技の威力 + 攻撃 - 防御) × タイプ相性 × タイプ一致 × 急所 × 乱数(0.85〜1.0)
func CalculateDamage(params CalculateDamageParams) DamageResult {
	if params.Power <= 0 {
		return DamageResult{Damage: 0, Critical: false}
	}

	modifiedAttack := GetModifiedStat(params.Attacker.Attack, params.Attacker.StatStages.Attack)
	attack := getBurnModifiedAttack(modifiedAttack, params.Attacker)
	defense := GetModifiedStat(params.Defender.Defense, params.Defender.StatStages.Defense)

	typeMultiplier := GetTypeMultiplier(params.MoveType, params.Defender.Type)
	stabMultiplier := getStabMultiplier(params.Attacker, params.MoveType)

	critical := isCriticalHit()
	criticalMultiplier := getCriticalMultiplier(critical)
	randomMultiplier := getRandomMultiplier()

	baseDamage := params.Power + attack - defense
	damage := maxInt(1, baseDamage)

	final := int(math.Floor(
		float64(damage) * typeMultiplier * stabMultiplier * criticalMultiplier * randomMultiplier,
	))

	return DamageResult{
		Damage:   maxInt(1, final),
		Critical: critical,
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
